File: ios/Jazzify/Main/FreeTierBlockLocks.swift
import Foundation

enum FreeTierBlockLocks {
    /// フリー会員向けに `maxBlockNumber` 超のブロックをロックする。
    static func apply(
        graph: LessonJourneyAccessGraph,
        lessons: [Lesson],
        maxBlockNumber: Int,
        isPremium: Bool
    ) -> LessonJourneyAccessGraph {
        guard !isPremium else { return graph }

        var lessonStates = graph.lessonStates
        var blockStates = graph.blockStates

        for lesson in lessons {
            let bn = lesson.blockNumber ?? 1
            guard bn > maxBlockNumber else { continue }
            guard let prev = lessonStates[lesson.id] else { continue }
            lessonStates[lesson.id] = LessonJourneyAccessGraph.LessonState(
                isUnlocked: false,
                isCompleted: prev.isCompleted
            )
        }

        for (bn, bs) in graph.blockStates where bn > maxBlockNumber {
            blockStates[bn] = LessonJourneyAccessGraph.BlockState(
                blockNumber: bn,
                isUnlocked: false,
                isCompleted: bs.isCompleted
            )
        }

        return LessonJourneyAccessGraph(lessonStates: lessonStates, blockStates: blockStates)
    }
}

File: src/utils/freeTierBlockLocks.ts
import type { Lesson } from '@/types';
import type { LessonAccessGraph } from '@/utils/lessonAccess';

/**
 * フリー会員向けに `maxBlockNumber` 超のブロックをロックする。
 * プレミアム会員のときは入力をそのまま返す（参照同一）。
 */
export function applyFreeTierBlockLocks(
  accessGraph: LessonAccessGraph,
  lessons: readonly Lesson[],
  maxBlockNumber: number,
  isPremiumMember: boolean,
): LessonAccessGraph {
  if (isPremiumMember) {
    return accessGraph;
  }

  const lessonStates = { ...accessGraph.lessonStates };
  const blockStates = { ...accessGraph.blockStates };

  for (const lesson of lessons) {
    const bn = lesson.block_number ?? 1;
    if (bn > maxBlockNumber) {
      const prev = lessonStates[lesson.id];
      if (prev !== undefined) {
        lessonStates[lesson.id] = { ...prev, isUnlocked: false };
      }
    }
  }

  for (const key of Object.keys(blockStates)) {
    const bn = Number(key);
    if (!Number.isFinite(bn) || bn <= maxBlockNumber) {
      continue;
    }
    const prev = blockStates[bn];
    if (prev !== undefined) {
      blockStates[bn] = { ...prev, isUnlocked: false };
    }
  }

  return { lessonStates, blockStates };
}
